src/SharpCompress/Utility.cs

```diff
@@ -277,8 +277,38 @@ public static long TransferTo(this Stream source, Stream destination)
             long total = 0;
             while (ReadTransferBlock(source, array, out var count))
             {
+                destination.Write(array, 0, count);
                 total += count;
+            }
+            return total;
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(array);
+        }
+    }
+
+    public static long TransferTo(this Stream source, Stream destination, long maxLength)
+    {
+        var array = GetTransferByteArray();
+        try
+        {
+            long total = 0;
+            var remaining = maxLength;
+            var maxReadSize = array.Length;
+            if (remaining < maxLength)
+            {
+                maxReadSize = (int)remaining;
+            }
+            while (ReadTransferBlock(source, array, maxReadSize, out var count))
+            {
                 destination.Write(array, 0, count);
+                total += count;
+                if (remaining - count < 0)
+                {
+                    break;
+                }
+                remaining -= count;
```

Comment:
missing maxReadSize adjustment. This reintroduces the bug that used to be there
if (remaining < maxReadSize)
{
    maxReadSize = (int)remaining;
}

Reply:
you're right